lib/engines/content_block_manager/app/models/content_block_manager/content_block/edition/has_audit_trail.rb

```diff
@@ -28,7 +28,43 @@ def record_update
       unless draft?
         user = Current.user
         state = try(:state)
-        versions.create!(event: "updated", user:, state:)
+        versions.create!(event: "updated", user:, state:, field_diff:)
+      end
+    end
+
+    FIELD_DIFF = Data.define(:previous_value, :new_value)
```

Comment:
I wonder if you might be better having `FIELD_DIFF` as a separate class with a `.all_for_document` class method. That way you can test it in isolation a bit better, as well as breaking down the methods in `field_diff` a bit more. Something like:

```
class FieldDiff < Data.define(:previous_value, :new_value)
  def self.all_for_document
     # logic from `field_diff` goes here, maybe broken down into smaller methods
  end
end
```

Reply:
I'm also wondering if you could have a way of translating the JSON into a `FieldDiff` object, that way if there are any changes to the structure, you could handle that programatically, rather than having to deal with JSON objects that may have changed over time

Reply:
good shout about breaking out the logic, will take a look.

> I'm also wondering if you could have a way of translating the JSON into a FieldDiff object, 
I was wondering if there was way to use jsonschema in some way - basically it should always follow this shape
```
{
  {field_key}: {
    "previous_value": string,
    "new_value": string,
  }
}
```
but with the `field_key` being unknown i don't how to say that in rails land 🤔

Reply:
Maybe having an array of objects like this might be more predictable:

```
[
  { "key": string, "previous_value": string, "new_value": string }
]
```

Reply:
I tried the array part in the last approach but it's a nightmare to then try and work out the order and display it correctly on the front end...

Reply:
though maybe now we are doing the ordering in the audit_trail the frontend is less a problem, as long as the order was preserved when we iterate through to make the rows...

Reply:
ok I've had a go at your suggestions and think it's looking better 👍